File: 1970.py
# leitura dos dados de entrada
n, k = map(int, input().split())
music_duration = list(map(int, input().split()))
max_recording = list(map(int, input().split()))
    
# preenche os cartuchos restantes com 0
while len(max_recording) < 3:
  max_recording.append(0)
    
# programação dinâmica com memoização
memo = {}
# idx: índice da música atual
# cap1, cap2, cap3: capacidades restantes dos cartuchos
def dp(idx, cap1, cap2, cap3):
  if idx == n:
    return 0
        
  state = (idx, cap1, cap2, cap3)
  if state in memo:
    return memo[state]
        
  # opção 1: não gravar esta música
  max_val = dp(idx + 1, cap1, cap2, cap3)
        
  # opção 2: tentar gravar em cada cartucho
  # obtem a duração da música atual
  duration = music_duration[idx]
  
  # tenta gravar no cartucho 1
  if cap1 >= duration:
      max_val = max(max_val, duration + dp(idx + 1, cap1 - duration, cap2, cap3))
  
  # tenta gravar no cartucho 2
  if cap2 >= duration:
      max_val = max(max_val, duration + dp(idx + 1, cap1, cap2 - duration, cap3))
  
  # tenta gravar no cartucho 3
  if cap3 >= duration:
      max_val = max(max_val, duration + dp(idx + 1, cap1, cap2, cap3 - duration))
  
  # armazena o resultado no memo
  memo[state] = max_val
  # retorna o valor máximo encontrado
  return max_val
    
result = dp(0, max_recording[0], max_recording[1], max_recording[2])
print(result)

File: 1970.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_NUM_MUSICS 100
#define MAX_DURATION 50
#define MAX_CARTRIFGE 3

int main() {
    int n, k, i, j;
    scanf("%d %d", &n, &k);

    // leitura da duração das músicas
    int music_duration[MAX_NUM_MUSICS];
    for (i = 0; i < n; i++)
    {
        scanf("%d", &music_duration[i]);
    }

    // leitura da duração dos cartuchos
    int max_recording[MAX_CARTRIFGE] = {0, 0, 0};
    for (i = 0; i < k; i++)
    {
        scanf("%d", &max_recording[i]);
    }

    // preenche os cartuchos restantes com 0
    for (int i = k; i < 3; i++) {
        max_recording[i] = 0;
    }

    int cartridge1 = max_recording[0];
    int cartridge2 = max_recording[1];
    int cartridge3 = max_recording[2];

    // ordena músicas em ordem decrescente (bubble sort)
    for (i = 0; i < n - 1; i++)
    {
        for (j = i + 1; j < n; j++)
        {
            if (music_duration[i] < music_duration[j])
            {
                int temp = music_duration[i];
                music_duration[i] = music_duration[j];
                music_duration[j] = temp;
            }
        }
    }

    // DP: dp[cap1][cap2][cap3] = máximo de minutos gravados
    int dp[MAX_DURATION][MAX_DURATION][MAX_DURATION];
    memset(dp, -1, sizeof(dp));
    dp[cartridge1][cartridge2][cartridge3] = 0;
    int max_minutes = 0;

    // para cada música
    for (int idx = 0; idx < n; idx++) {
        int duration = music_duration[idx];
        int new_dp[MAX_DURATION][MAX_DURATION][MAX_DURATION];
        memcpy(new_dp, dp, sizeof(dp));

        for (int rem1 = 0; rem1 <= cartridge1; rem1++)
        {
            for (int rem2 = 0; rem2 <= cartridge2; rem2++)
            {
                for (int rem3 = 0; rem3 <= cartridge3; rem3++)
                {
                    if (dp[rem1][rem2][rem3] != -1) {
                        int current = dp[rem1][rem2][rem3];
                        if (current > max_minutes)
                        {
                            max_minutes = current;
                        }

                        if (rem1 >= duration) {
                            int new_val = current + duration;
                            if (new_val > new_dp[rem1 - duration][rem2][rem3]) {
                                new_dp[rem1 - duration][rem2][rem3] = new_val;
                            }
                        }
                        
                        if (rem2 >= duration) {
                            int new_val = current + duration;
                            if (new_val > new_dp[rem1][rem2 - duration][rem3]) {
                                new_dp[rem1][rem2 - duration][rem3] = new_val;
                            }
                        }
                        
                        if (rem3 >= duration) {
                            int new_val = current + duration;
                            if (new_val > new_dp[rem1][rem2][rem3 - duration]) {
                                new_dp[rem1][rem2][rem3 - duration] = new_val;
                            }
                        }
                    }
                }
            }
        }
        memcpy(dp, new_dp, sizeof(dp));
    }
    
    // verifica o estado final
    for (int rem1 = 0; rem1 <= cartridge1; rem1++)
    {
        for (int rem2 = 0; rem2 <= cartridge2; rem2++)
        {
            for (int rem3 = 0; rem3 <= cartridge3; rem3++)
            {
                if (dp[rem1][rem2][rem3] > max_minutes) {
                    max_minutes = dp[rem1][rem2][rem3];
                }
            }
        }
    }

    printf("%d\n", max_minutes);
    return 0;
}